<a href="https://colab.research.google.com/github/Anna-Shatalova/STUDY-Data_Analysis/blob/main/A_Shatalova_HW_01_pandas_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание 1
*Выполнила: Шаталова Анна* ✌

Все задания ниже ~коммунисты~ имеют равный вес.

## Описание данных

В папке Data находится информация о студентах. Всего 10 групп студентов.  
Файлы делятся на две категории:
*   `Students_info_i` - информация о студентах из группы `i`
*   `Students_marks_i` - оценки студентов из группы `i` за экзамены



Одно из важных достоинств pandas $-$ это удобные методы реляционного взаимодействия с данными, аналогичные, например, возможностям SQL для слияния и конкатенации таблиц: merge, join, concat. Наличие готовых методов позволяет не реализовывать самостоятельно поэлементную обработку данных и оперировать сразу целыми таблицами данных.

Подробно об этих методах посмотрите тут:

 https://www.kaggle.com/residentmario/renaming-and-combining#Combining

## Практика

In [2]:
# импортируем библиотеку Pandas
import pandas as pd

In [3]:
%%capture
# загружаем данные
!wget https://github.com/evgpat/datasets/raw/main/students.zip

In [4]:
# проверяем загрузку (наличие файла students.zip)
!ls -lah

total 32K
drwxr-xr-x 1 root root 4.0K Mar  3 11:18 .
drwxr-xr-x 1 root root 4.0K Mar  3 11:18 ..
drwxr-xr-x 4 root root 4.0K Feb 28 14:26 .config
drwxr-xr-x 1 root root 4.0K Feb 28 14:27 sample_data
-rw-r--r-- 1 root root  16K Mar  3 11:18 students.zip


In [5]:
# разархивируем данные
!unzip students.zip

Archive:  students.zip
   creating: students/
  inflating: students/Students_info_0.csv  
  inflating: students/Students_info_1.csv  
  inflating: students/Students_info_2.csv  
  inflating: students/Students_info_3.csv  
  inflating: students/Students_info_4.csv  
  inflating: students/Students_info_5.csv  
  inflating: students/Students_info_6.csv  
  inflating: students/Students_info_7.csv  
  inflating: students/Students_info_8.csv  
  inflating: students/Students_info_9.csv  
  inflating: students/Students_marks_0.csv  
  inflating: students/Students_marks_1.csv  
  inflating: students/Students_marks_2.csv  
  inflating: students/Students_marks_3.csv  
  inflating: students/Students_marks_4.csv  
  inflating: students/Students_marks_5.csv  
  inflating: students/Students_marks_6.csv  
  inflating: students/Students_marks_7.csv  
  inflating: students/Students_marks_8.csv  
  inflating: students/Students_marks_9.csv  


In [6]:
# смотрим что получилось
!ls -lah students

total 128K
drwxrwxr-x 2 root root 4.0K Jul 25  2023 .
drwxr-xr-x 1 root root 4.0K Mar  3 11:18 ..
-rw-rw-r-- 1 root root 5.8K Jan 17  2021 Students_info_0.csv
-rw-rw-r-- 1 root root 5.9K Jan 17  2021 Students_info_1.csv
-rw-rw-r-- 1 root root 5.9K Jan 17  2021 Students_info_2.csv
-rw-rw-r-- 1 root root 5.9K Jan 17  2021 Students_info_3.csv
-rw-rw-r-- 1 root root 5.9K Jan 17  2021 Students_info_4.csv
-rw-rw-r-- 1 root root 5.9K Jan 17  2021 Students_info_5.csv
-rw-rw-r-- 1 root root 5.9K Jan 17  2021 Students_info_6.csv
-rw-rw-r-- 1 root root 5.9K Jan 17  2021 Students_info_7.csv
-rw-rw-r-- 1 root root 5.9K Jan 17  2021 Students_info_8.csv
-rw-rw-r-- 1 root root 6.0K Jan 17  2021 Students_info_9.csv
-rw-rw-r-- 1 root root 1.4K Jan 17  2021 Students_marks_0.csv
-rw-rw-r-- 1 root root 1.5K Jan 17  2021 Students_marks_1.csv
-rw-rw-r-- 1 root root 1.5K Jan 17  2021 Students_marks_2.csv
-rw-rw-r-- 1 root root 1.5K Jan 17  2021 Students_marks_3.csv
-rw-rw-r-- 1 root root 1.5K Jan 17  2021 Stu

### ◼ Задание 1

Соберите всю информацию о студентах в одну таблицу df. В получившейся таблице должна быть информация и оценки всех студентов из всех групп. Напечатайте несколько строк таблицы для демонстрации результата

⚡ *c помощью модуля Python glob можно искать все имена путей, которые находят файлы, соответствующие определенному шаблону*

In [7]:
from glob import glob

✅ Собираем все таблицы с названием **students_info** в одну

In [8]:
files_1 = glob('/content/students/Students_info*')

In [9]:
# поскольку модуль glob возвращает все в произвольном порядке отсортируем результат
# все файлы на месте
files_1.sort()
files_1

['/content/students/Students_info_0.csv',
 '/content/students/Students_info_1.csv',
 '/content/students/Students_info_2.csv',
 '/content/students/Students_info_3.csv',
 '/content/students/Students_info_4.csv',
 '/content/students/Students_info_5.csv',
 '/content/students/Students_info_6.csv',
 '/content/students/Students_info_7.csv',
 '/content/students/Students_info_8.csv',
 '/content/students/Students_info_9.csv']

In [10]:
# делаем все четко, как говорит индус из видео на ютубе
# не очень понятно что именно, но главное - работает!
df_1 = pd.concat((pd.read_csv(file) for file in files_1), ignore_index = True)

In [11]:
df_1.shape

(1000, 7)

In [12]:
df_1.head()

index  gender race/ethnicity parental level of education         lunch  \
0      0  female        group B           bachelor's degree      standard   
1      1  female        group C                some college      standard   
2      2  female        group B             master's degree      standard   
3      3    male        group A          associate's degree  free/reduced   
4      4    male        group C                some college      standard   

  test preparation course   group  
0                    none  group1  
1               completed  group1  
2                    none  group1  
3                    none  group1  
4                    none  group1

✅ Повторяем все, что делали выше, но для таблиц с оценками студентов - **students_marks**

In [13]:
files_2 = glob('/content/students/Students_marks*')

In [14]:
files_2.sort()
files_2

['/content/students/Students_marks_0.csv',
 '/content/students/Students_marks_1.csv',
 '/content/students/Students_marks_2.csv',
 '/content/students/Students_marks_3.csv',
 '/content/students/Students_marks_4.csv',
 '/content/students/Students_marks_5.csv',
 '/content/students/Students_marks_6.csv',
 '/content/students/Students_marks_7.csv',
 '/content/students/Students_marks_8.csv',
 '/content/students/Students_marks_9.csv']

In [15]:
df_2 = pd.concat((pd.read_csv(file) for file in files_2), ignore_index = True)

In [16]:
df_2.shape

(1000, 4)

In [17]:
df_2.head()

index  math score  reading score  writing score
0      0          72             72             74
1      1          69             90             88
2      2          90             95             93
3      3          47             57             44
4      4          76             78             75

✅ Объединяем обе таблицы ~остаться должна только одна~

In [18]:
df = pd.merge(df_1, df_2, left_index = True, right_index = True)
df.shape

(1000, 11)

In [19]:
df.head(10)

index_x  gender race/ethnicity parental level of education         lunch  \
0        0  female        group B           bachelor's degree      standard   
1        1  female        group C                some college      standard   
2        2  female        group B             master's degree      standard   
3        3    male        group A          associate's degree  free/reduced   
4        4    male        group C                some college      standard   
5        5  female        group B          associate's degree      standard   
6        6  female        group B                some college      standard   
7        7    male        group B                some college  free/reduced   
8        8    male        group D                 high school  free/reduced   
9        9  female        group B                 high school  free/reduced   

  test preparation course   group  index_y  math score  reading score  \
0                    none  group1        0          72             72   
1               completed  group1        1          69             90   
2                    none  group1        2          90             95   
3                    none  group1        3          47             57   
4                    none  group1        4          76             78   
5                    none  group1        5          71             83   
6               completed  group1        6          88             95   
7                    none  group1        7          40             43   
8               completed  group1        8          64             64   
9                    none  group1        9          38             60   

   writing score  
0             74  
1             88  
2             93  
3             44  
4             75  
5             78  
6             92  
7             39  
8             67  
9             50

### ◼ Задание 2

Удалите столбец index у полученной таблицы. Напечатайте первые 10 строк таблицы

In [20]:
# запускать один раз и с сторожностью (с) Метод Проб и Ошибок
df.drop(['index_x', 'index_y'], axis=1 , inplace=True)

In [21]:
df.head(10)

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   
7    male        group B                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test preparation course   group  math score  reading score  writing score  
0                    none  group1          72             72             74  
1               completed  group1          69             90             88  
2                    none  group1          90             95             93  
3                    none  group1          47             57             44  
4                    none  group1          76             78             75  
5                    none  group1          71             83             78  
6               completed  group1          88             95             92  
7                    none  group1          40             43             39  
8               completed  group1          64             64             67  
9                    none  group1          38             60             50

### ◼ Задание 3

Выведите на экран размеры полученной таблицы

In [22]:
df.shape

(1000, 9)

✅ Можно по-идее еще вот так сделать:

In [23]:
df.size

9000

In [24]:
len(df)

1000

### ◼ Задание 4

Выведите на экран статистические характеристики числовых столбцов таблицы (минимум, максимум, среднее значение, стандартное отклонение)

✅ Максимум

In [25]:
maximums = df.max()
print(maximums)

gender                                     male
race/ethnicity                          group E
parental level of education    some high school
lunch                                  standard
test preparation course                    none
group                                    group9
math score                                  100
reading score                               100
writing score                               100
dtype: object


In [26]:
df.max()

gender                                     male
race/ethnicity                          group E
parental level of education    some high school
lunch                                  standard
test preparation course                    none
group                                    group9
math score                                  100
reading score                               100
writing score                               100
dtype: object

✅ Минимум

In [27]:
minimums = df.min()
print(minimums)

gender                                     female
race/ethnicity                            group A
parental level of education    associate's degree
lunch                                free/reduced
test preparation course                 completed
group                                      group1
math score                                      0
reading score                                  17
writing score                                  10
dtype: object


In [28]:
df.min()

gender                                     female
race/ethnicity                            group A
parental level of education    associate's degree
lunch                                free/reduced
test preparation course                 completed
group                                      group1
math score                                      0
reading score                                  17
writing score                                  10
dtype: object

✅ Стандартное отклонение

In [29]:
df.std()

<ipython-input-29-ce97bb7eaef8>:1: FutureWarning: The default value of numeric_only in DataFrame.std is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.std()


math score       15.163080
reading score    14.600192
writing score    15.195657
dtype: float64

In [30]:
# чтобы не было "ругани" - берем только числовые столбцы
df[['math score', 'reading score', 'writing score']].std()

math score       15.163080
reading score    14.600192
writing score    15.195657
dtype: float64

✅ Среднее значение

In [31]:
df[['math score', 'reading score', 'writing score']].mean()

math score       66.089
reading score    69.169
writing score    68.054
dtype: float64

### ◼ Задание 5

Проверьте, есть ли в таблице пропущенные значения

✅ Выглядит так, словно бы все на месте, без пропусков

In [32]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   group                        1000 non-null   object
 6   math score                   1000 non-null   int64 
 7   reading score                1000 non-null   int64 
 8   writing score                1000 non-null   int64 
dtypes: int64(3), object(6)
memory usage: 70.4+ KB
None


### ◼ Задание 6

Выведите на экран средние баллы студентов по каждому предмету (`math`, `reading`, `writing`)

In [33]:
df[['math score', 'reading score', 'writing score']].mean()

math score       66.089
reading score    69.169
writing score    68.054
dtype: float64

### 🤖 Задание 7

Как зависят оценки от того, проходил ли студент курс для подготовки к сдаче экзамена (test preparation course)?

Выведите на экран для каждого предмета в отдельности средний балл студентов, проходивших курс для подготовки к экзамену и не проходивших курс.

In [64]:
# спонсор решения - ChatBox GPT
average_scores = df.groupby('test preparation')[['math score', 'reading score', 'writing score']].mean()
print(average_scores)

                  math score  reading score  writing score
test preparation                                          
completed          69.695531      73.893855      74.418994
none               64.077882      66.534268      64.504673


✅ **Вывод:** прохождение курса подготовки к сдаче экзамена повышает средний балл. Наибольшее улучшение, судя по всему, идет по письменному экзмену и незначительное - по математике ~рожденный ползать считать не может~.

### ◼ Задание 8

Выведите на экран все различные значения из столбца `lunch`

In [34]:
df['lunch'].unique()

array(['standard', 'free/reduced'], dtype=object)

In [35]:
# так тоже работает
df.lunch.unique()

array(['standard', 'free/reduced'], dtype=object)

In [36]:
# считает количество уникальных в указанном столбце
df['lunch'].value_counts()

standard        645
free/reduced    355
Name: lunch, dtype: int64

### ◼ Задание 9

Переименуйте колонку `parental level of education` в `education`, а `test preparation course` в `test preparation` с помощью метода pandas [`rename`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html)

In [37]:
df = df.rename(columns={'parental level of education': 'education', 'test preparation course': 'test preparation'})
df.head()

gender race/ethnicity           education         lunch test preparation  \
0  female        group B   bachelor's degree      standard             none   
1  female        group C        some college      standard        completed   
2  female        group B     master's degree      standard             none   
3    male        group A  associate's degree  free/reduced             none   
4    male        group C        some college      standard             none   

    group  math score  reading score  writing score  
0  group1          72             72             74  
1  group1          69             90             88  
2  group1          90             95             93  
3  group1          47             57             44  
4  group1          76             78             75

### ◼ Задание 10

**Зафиксируем минимальный балл для сдачи экзамена**

In [38]:
passmark = 50

**Ответьте на вопросы**
* Какая доля студентов сдала экзамен по математике (passmark > 50)?
* Какая доля студентов, проходивших курс подготовки к экзамену, сдала экзамен по математике?
* Какая доля женщин, не проходивших курс подготовки к экзамену, не сдала экзамен по математике?

In [39]:
# студенты, сдавшие экзамен по математике (счастливчики)
len(df[(df["math score"] > passmark) & (df.gender != "none")])/len(df[(df.gender != "none")])

0.85

In [40]:
# студенты, сдавшие экземен по математике, которые ходили на подготовительные курсы (мажоры)
len(df[(df["test preparation"] != "none") & (df["math score"] > passmark) & (df.gender != "none")])/len(df[(df["test preparation"] != "none") & (df.gender != "none")])

0.9078212290502793

In [41]:
# девушки, не сдавшие математику и не прошедшие подготовительные курсы (гуманитарии)
len(df[(df["test preparation"] == "none") & (df["math score"] < passmark) & (df.gender == "female")])/len(df[(df["test preparation"] == "none") & (df.gender == "female")])

0.20958083832335328

### ◼ Задание 11

С помощью `groupby` выполните задания ниже. Также выведите время выполнения каждого из заданий.
* Для каждой этнической группы выведите средний балл за экзамен по чтению
* Для каждого уровня образования выведите минимальный балл за экзамен по письму

✅ Подзадача 11.1 (этнические группы и средний бвлл по чтению)

In [42]:
import time

In [43]:
start_time = time.time() # время старта выполнения

# применяем встроенную математическую функцию агрегирования (среднее значение) к экзамену по чтению
agg_func_math = {
    'reading score': ['mean']
}
a = df.groupby(['race/ethnicity']).agg(agg_func_math).round(2)
print(a)

end_time = time.time()  # время окончания выполнения
execution_time = end_time - start_time  # вычисляем время выполнения

print(f"Время выполнения подзадачи 11.1: {execution_time} секунд")

               reading score
                        mean
race/ethnicity              
group A                64.67
group B                67.35
group C                69.10
group D                70.03
group E                73.03
Время выполнения подзадачи 11.1: 0.022050142288208008 секунд


✅ Подзадача 2 (уровень образования и минимальный балл по письму)

In [44]:
start_time = time.time()

# применяем встроенную математическую функцию агрегирования (минимальное значение) к экзамену по письму
agg_func_math = {
    'writing score': ['min']
}
b = df.groupby(['education']).agg(agg_func_math).round(2)
print(b)

end_time = time.time()
execution_time = end_time - start_time

print(f"Время выполнения подзадачи 11.2: {execution_time} секунд")

                   writing score
                             min
education                       
associate's degree            35
bachelor's degree             38
high school                   15
master's degree               46
some college                  19
some high school              10
Время выполнения подзадачи 11.2: 0.019149303436279297 секунд


### 🤖 Задание 12

Выполните задание 11 с помощью циклов. Сравните время выполнения

In [131]:
# Задание 1: Для каждой этнической группы выведите средний балл за экзамен по чтению
start_time = time.time()

for group, group_df in df.groupby('race/ethnicity'):
  avg_reading_score = group_df['reading score'].mean()
  print(f"Средний балл за экзамен по чтению для этнической группы {group}: {avg_reading_score}")
  print("Время выполнения задания 1:", time.time() - start_time)

Средний балл за экзамен по чтению для этнической группы group A: 64.67415730337079
Время выполнения задания 1: 0.005011796951293945
Средний балл за экзамен по чтению для этнической группы group B: 67.35263157894737
Время выполнения задания 1: 0.008109331130981445
Средний балл за экзамен по чтению для этнической группы group C: 69.10344827586206
Время выполнения задания 1: 0.009087324142456055
Средний балл за экзамен по чтению для этнической группы group D: 70.03053435114504
Время выполнения задания 1: 0.009466409683227539
Средний балл за экзамен по чтению для этнической группы group E: 73.02857142857142
Время выполнения задания 1: 0.009774208068847656


In [133]:
# Задание 2: Для каждого уровня образования выведите минимальный балл за экзамен по письму
start_time = time.time()

for edu_level, edu_df in df.groupby('education'):
  min_writing_score = edu_df['writing score'].min()
  print(f"Минимальный балл за экзамен по письму для уровня образования '{edu_level}': {min_writing_score}")
  print("Время выполнения задания 2:", time.time() - start_time)

Минимальный балл за экзамен по письму для уровня образования 'associate's degree': 35
Время выполнения задания 2: 0.004056453704833984
Минимальный балл за экзамен по письму для уровня образования 'bachelor's degree': 38
Время выполнения задания 2: 0.00952458381652832
Минимальный балл за экзамен по письму для уровня образования 'high school': 15
Время выполнения задания 2: 0.010524988174438477
Минимальный балл за экзамен по письму для уровня образования 'master's degree': 46
Время выполнения задания 2: 0.010863780975341797
Минимальный балл за экзамен по письму для уровня образования 'some college': 19
Время выполнения задания 2: 0.014222860336303711
Минимальный балл за экзамен по письму для уровня образования 'some high school': 10
Время выполнения задания 2: 0.014864206314086914


### ◼ Задание 13

Выведите на экран средние баллы студентов по каждому предмету в зависимости от пола и уровня образования.  
То есть должно получиться количество групп, равных 2 * (число уровней образования), и для каждой такой группы выведите средний балл по каждому из предметов.

Это можно сделать с помощью сводных таблиц ([pivot_table](https://www.kaggle.com/kamilpolak/tutorial-how-to-use-pivot-table-in-pandas))

In [46]:
# судя по статье по ссылке нам понадобиться Numpy
import numpy as np

In [47]:
# сначала указываем пол и образование это два первых столбца
# считаем средний балл по каждому предмету
grouped = pd.pivot_table(df,index=['gender','education'],aggfunc={'math score':np.mean,'reading score':np.mean, 'writing score': np.mean})
grouped

math score  reading score  writing score
gender education                                                   
female associate's degree   65.250000      74.120690      74.000000
       bachelor's degree    68.349206      77.285714      78.380952
       high school          59.351064      68.202128      66.691489
       master's degree      66.500000      76.805556      77.638889
       some college         65.406780      73.550847      74.050847
       some high school     59.296703      69.109890      68.285714
male   associate's degree   70.764151      67.433962      65.405660
       bachelor's degree    70.581818      68.090909      67.654545
       high school          64.705882      61.480392      58.539216
       master's degree      74.826087      73.130435      72.608696
       some college         69.009259      64.990741      63.148148
       some high school     67.840909      64.693182      61.375000

### ◼ Задание 14

Сколько студентов успешно сдали экзамен по математике?

Создайте новый столбец в таблице `df` под названием `Math_PassStatus` и запишите в него F, если студент не сдал экзамен по математике (балл за экзамен < passmark), и P иначе.

Посчитайте количество студентов, сдавших и не сдавших экзамен по математике.

Сделайте аналогичные шаги для экзаменов по чтению и письму.

✅ Экзамен по математике:

In [48]:
# создаем новый столбец и заносим в него F и P в зависимости от факта сдачи или не сдачи экзамена по математике
df['Math_PassStatus'] = np.where(df['math score'] > passmark, 'P', 'F')
df

gender race/ethnicity           education         lunch test preparation  \
0    female        group B   bachelor's degree      standard             none   
1    female        group C        some college      standard        completed   
2    female        group B     master's degree      standard             none   
3      male        group A  associate's degree  free/reduced             none   
4      male        group C        some college      standard             none   
..      ...            ...                 ...           ...              ...   
995  female        group E     master's degree      standard        completed   
996    male        group C         high school  free/reduced             none   
997  female        group C         high school  free/reduced        completed   
998  female        group D        some college      standard        completed   
999  female        group D        some college  free/reduced             none   

       group  math score  reading score  writing score Math_PassStatus  
0     group1          72             72             74               P  
1     group1          69             90             88               P  
2     group1          90             95             93               P  
3     group1          47             57             44               F  
4     group1          76             78             75               P  
..       ...         ...            ...            ...             ...  
995  group10          88             99             95               P  
996  group10          62             55             55               P  
997  group10          59             71             65               P  
998  group10          68             78             77               P  
999  group10          77             86             86               P  

[1000 rows x 10 columns]

In [104]:
# считаем сколько студентов успешно сдали математику, т.е. имеют значение P в столбце Math_PassStatus
len(df[df['Math_PassStatus'] =='P'])

850

In [105]:
# количество не математиков:
len(df[df['Math_PassStatus'] =='F'])

150

✅ Экзамен по чтению:

In [49]:
df['Reading_PassStatus'] = np.where(df['reading score'] > passmark, 'P', 'F')
df

gender race/ethnicity           education         lunch test preparation  \
0    female        group B   bachelor's degree      standard             none   
1    female        group C        some college      standard        completed   
2    female        group B     master's degree      standard             none   
3      male        group A  associate's degree  free/reduced             none   
4      male        group C        some college      standard             none   
..      ...            ...                 ...           ...              ...   
995  female        group E     master's degree      standard        completed   
996    male        group C         high school  free/reduced             none   
997  female        group C         high school  free/reduced        completed   
998  female        group D        some college      standard        completed   
999  female        group D        some college  free/reduced             none   

       group  math score  reading score  writing score Math_PassStatus  \
0     group1          72             72             74               P   
1     group1          69             90             88               P   
2     group1          90             95             93               P   
3     group1          47             57             44               F   
4     group1          76             78             75               P   
..       ...         ...            ...            ...             ...   
995  group10          88             99             95               P   
996  group10          62             55             55               P   
997  group10          59             71             65               P   
998  group10          68             78             77               P   
999  group10          77             86             86               P   

    Reading_PassStatus  
0                    P  
1                    P  
2                    P  
3                    P  
4                    P  
..                 ...  
995                  P  
996                  P  
997                  P  
998                  P  
999                  P  

[1000 rows x 11 columns]

In [106]:
len(df[df['Reading_PassStatus'] =='P'])

903

In [107]:
len(df[df['Reading_PassStatus'] =='F'])

97

✅ Письменный экзамен:

In [72]:
df['Writing_PassStatus'] = np.where(df['writing score'] > passmark, 'P', 'F')
df

gender race/ethnicity           education         lunch test preparation  \
0    female        group B   bachelor's degree      standard             none   
1    female        group C        some college      standard        completed   
2    female        group B     master's degree      standard             none   
3      male        group A  associate's degree  free/reduced             none   
4      male        group C        some college      standard             none   
..      ...            ...                 ...           ...              ...   
995  female        group E     master's degree      standard        completed   
996    male        group C         high school  free/reduced             none   
997  female        group C         high school  free/reduced        completed   
998  female        group D        some college      standard        completed   
999  female        group D        some college  free/reduced             none   

       group  math score  reading score  writing score Math_PassStatus  \
0     group1          72             72             74               P   
1     group1          69             90             88               P   
2     group1          90             95             93               P   
3     group1          47             57             44               F   
4     group1          76             78             75               P   
..       ...         ...            ...            ...             ...   
995  group10          88             99             95               P   
996  group10          62             55             55               P   
997  group10          59             71             65               P   
998  group10          68             78             77               P   
999  group10          77             86             86               P   

    Reading_PassStatus Writing_PassStatus  
0                    P                  P  
1                    P                  P  
2                    P                  P  
3                    P                  F  
4                    P                  P  
..                 ...                ...  
995                  P                  P  
996                  P                  P  
997                  P                  P  
998                  P                  P  
999                  P                  P  

[1000 rows x 12 columns]

In [108]:
len(df[df['Writing_PassStatus'] =='P'])

876

In [109]:
len(df[df['Writing_PassStatus'] =='F'])

124

✅ **ВЫВОД:**

Экземен по математике успешно сдали  850 студентов (завалили 150)

Экзамен по письму успешно сдали 876 студентов (завалили 124)

Экземан по чтению успешно сдали 903 студента (завалили 97)

In [51]:
# df.drop(['название столбца'], axis=1 , inplace=True) - использовать при необходимости удаления лишних столбцов

### 🤖 Задание 15

Сколько студентов успешно сдали все экзамены?

Создайте столбец `OverAll_PassStatus` и запишите в него для каждого студента 'F', если студент не сдал хотя бы один из трех экзаменов, а иначе 'P'.

Посчитайте количество студентов, которые сдали все экзамены.

✅ Предварительно по тем данным, которые уже есть считаем сколько студентов сдали все экзамены, т.е. у них оценки по каждому из предметов более 50 (значение равно P).

In [52]:
len(df[(df["Math_PassStatus"] == 'P') & (df["Reading_PassStatus"] == 'P') & (df["Writing_PassStatus"] == 'P')])

796

✅ Узнали, что из 1000 студентов 796 справились со всеми экзаменами.

Проверим.

In [ ]:
# создаем новый столбец в таблице с названием 'OverAll_PassStatus'
df['OverAll_PassStatus'] = np.where((df['Math_PassStatus'] != 'P') | (df['Reading_PassStatus'] != 'P') | (df['Writing_PassStatus'] != 'P'), 'F', 'P')

In [122]:
# Подсчет количества студентов, успешно сдавших все три экзамена
pass_count = df['OverAll_PassStatus'].value_counts()['P']
print("\nКоличество студентов, успешно сдавших все три экзамена:", pass_count)


Количество студентов, успешно сдавших все три экзамена: 796


### 🤖 Задание 16

Переведем баллы в оценки

**Система перевода баллов в оценки**
*   [90;100] = A
*   [80;90) = B
*   [70;80) = C
*   [60;70) = D
*   [50;60) = E
*   [0;50) = F (Fail)

Создайте вспомогательную функцию, которая будет по среднему баллу за три экзамена выставлять оценку студенту по данным выше критериям.

Создайте столбец `Grade` и запишите в него оценку каждого студента.

Выведите количество студентов, получивших каждую из оценок.

In [103]:
# пишем вспомогательную функцию, которая определяет какому значению соответствует результат экзамена студента (А, B, C, D или F)
def GetGrade(average_mark):
    if 90 <= average_mark <= 100:
      return 'A'
    elif 80 <= average_mark < 90:
      return 'B'
    elif 70 <= average_mark < 80:
      return 'C'
    elif 60 <= average_mark < 70:
      return 'D'
    else:
      return 'F'

In [100]:
# создаем столбец, где находим среднее значение студента за 3 экзамена
# к этому новому столбцу применяем ранее написанную функцию, которая "переводит число в букву"
df['Grade'] = (df['math score'] + df['reading score'] + df['writing score']) / 3
df['Grade'] = df['Grade'].apply(GetGrade)

In [127]:
# считаем сколько значений по каждому показателю
grade_counts = df['Grade'].value_counts()

In [128]:
print("\nКоличество студентов, получивших каждую из оценок:")
print(grade_counts)


Количество студентов, получивших каждую из оценок:
F    285
C    261
D    256
B    146
A     52
Name: Grade, dtype: int64
